data_structures/treap.cpp

```diff
@@ -0,0 +1,248 @@
+/**
+ * @file
+ * @brief A balanced binary search tree (BST) on the basis of binary search tree
+ * and heap: the [Treap](https://en.wikipedia.org/wiki/Treap) algorithm
+ * implementation
+ *
+ * @details
+ * Implementation of the treap data structre
+ *
+ * Support operations including insert, erase, and query (the rank of specified
+ * element or the element ranked x) as the same as BST
+ *
+ * But these operations take O(log N) time, since treap keeps property of heap
+ * using rotate operation, and the desired depth of the tree is O(log N).
+ * There's very little chance that it will degenerate into a chain like BST
+ *
+ * @author [Kairao ZHENG](https://github.com/fgmn)
+ */
+
+#include <array>     /// For array
+#include <cassert>   /// For assert
+#include <iostream>  /// For IO operations
+
+/**
+ * @namespace
+ * @brief Data Structures
+ */
+namespace data_structures {
+/**
+ * @namespace
+ * @brief Functions for the [Treap](https://en.wikipedia.org/wiki/Treap)
+ * algorithm implementation
+ */
+namespace treap {
+const int maxNode = 1e6 + 5;  ///< maximum number of nodes
+/**
+ * @brief Struct representation of the treap
+ */
+struct Treap {
+    int root = 0;      ///< root of the treap
+    int treapCnt = 0;  ///< Total number of current nodes in the treap
+    std::array<int, maxNode> key;        ///< Node identifier
+    std::array<int, maxNode> priority;   ///< Random priority
+    std::array<int, 2> childs[maxNode];  ///< [i][0] represents the
```

Comment:
Are you trying to create a matrix here?
If so, you should use two-dimensional (2D) array (see [this article](https://cpptruths.blogspot.com/2011/10/multi-dimensional-arrays-in-c11.html) for more information).

```suggestion
    std::array<std::array<int, 2>, 2> childs = { { } };  ///< [i][0] represents the
```